<a id='top'></a>

# CSCI 3022: Intro to Data Science - Spring 2019 Practicum 2
***

This practicum is due on Canvas by **11:59 PM on Friday May 3**. Your solutions to theoretical questions should be done in Markdown/MathJax directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  

**Here are the rules:** 

1. All work, code and analysis, must be your own. 
1. You may use your course notes, posted lecture slides, textbooks, in-class notebooks, and homework solutions as resources.  You may also search online for answers to general knowledge questions like the form of a probability distribution function or how to perform a particular operation in Python/Pandas. 
1. This is meant to be like a coding portion of your midterm exam. So, the instructional team will be much less helpful than we typically are with homework. For example, we will not check answers, help debug your code, and so on.
1. If something is left open-ended, it is because we want to see how you approach the kinds of problems you will encounter in the wild, where it will not always be clear what sort of tests/methods should be applied. Feel free to ask clarifying questions though.
2. You may **NOT** post to message boards or other online resources asking for help.
3. You may **NOT** copy-paste solutions *from anywhere*.
4. You may **NOT** collaborate with classmates or anyone else.
5. In short, **your work must be your own**. It really is that simple.

Violation of the above rules will result in an immediate academic sanction (*at the very least*, you will receive a 0 on this practicum or an F in the course, depending on severity), and a trip to the Honor Code Council.

**By submitting this assignment, you agree to abide by the rules given above.**

***

**Name**:  Chakrya Ros

***


**NOTES**: 

- You may not use late days on the practicums nor can you drop your practicum grades. 
- If you have a question for us, post it as a **PRIVATE** message on Piazza.  If we decide that the question is appropriate for the entire class, then we will add it to a Practicum clarifications thread. 
- Do **NOT** load or use any Python packages that are not available in Anaconda 3.6. 
- Some problems with code may be autograded.  If we provide a function API **do not** change it.  If we do not provide a function API then you're free to structure your code however you like. 
- Submit only this Jupyter notebook to Canvas.  Do not compress it using tar, rar, zip, etc. 
- This should go without saying, but... For any question that asks you to calculate something, you **must show all work to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit.

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Bottom](#bot)

---

In [1]:
# import whatever packages you think you will need here!
from scipy import stats
import numpy as np 
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
%matplotlib inline


<br>

---
<a id='p1'></a>
[Back to top](#top)

### [30 points] Problem 1:  What's the best Oreo flavor?

> <img src="https://media2.s-nbcnews.com/j/newscms/2015_52/1351976/151223-watermelon-oreos-flickr-yh-1234p_a4c4ba1016261e5d799a789666948a6e.fit-760w.jpg" style="width: 400px;"/>
**Figure 1.** Watermelon Oreos. These are undeniably the absolute _worst_ Oreo flavor, but not a part of our data set. If they were, the problem would be too easy.

You're skipping down the road one day, singing a merry tune without a care in the world. Every so often you hop and click your heels together. After one such hop, though, you stumble and nearly trip. Under your feet is a fun data set!

In particular, this rascally data set includes ratings from 5 different individuals for 12 different types of Oreo cookie. Because some of the raters are allergic to disgusting flavors of Oreo, not every taster tasted every type of cookie. More information about the data set can be found [here](https://www.kaggle.com/rtatman/oreo-flavors-tastetest-ratings).

As you examine the data, a booming voice from the sky instructs you to analyze the data to answer two key questions:
1. is there some difference in mean rating given to different flavors of Oreo? and
2. which flavor (out of those in the data set) is the best?

**Part A:** Load the data set as a Pandas DataFrame. Then, process it as you see fit. For the analysis, you might create a new DataFrame, or other kinds of Python objects to use the data. Do what feels right, and include an **explanation** in Markdown of how you are processing the data. You **may not** alter the original data file.

**Solution:**

I load the data set as a Pandas DataFrame. Then I copy to new DataFrame. First, I replace the "-" with np.nan. 

In [2]:
df = pd.read_csv("oreo_rankings.csv")
dfnew = df.copy()
dfnew = dfnew.replace({"-":np.nan})
# dfnew.drop(columns=["notes_and_discussion"], inplace=True)
dfnew

oreo_flavor rater_n  rater_c  rater_r rater_k rater_a  \
0                   pb&j     1.5      4.0      2.0     NaN     NaN   
1              mega stuf     2.5      3.0      4.0     NaN     1.5   
2                  lemon     4.5      3.5      1.5     NaN     3.5   
3              chocolate     3.5      3.0      3.5       3       4   
4          birthday cake     2.5      5.0      2.5       4       3   
5            double stuf     3.5      4.0      3.0       3     3.5   
6             red velvet       4      3.0      4.5       4     2.5   
7   Dunkin' Donuts Mocha     2.5      1.0      5.0       5       5   
8                   mini       3      4.0      4.0     NaN    4.25   
9         thins, coconut     3.5      4.5      3.0     4.5     4.5   
10                  mint     3.5      4.0      4.5       4    4.25   
11          cinnamon bun     NaN      4.5      4.5     4.5       4   

    average_across_raters                               notes_and_discussion  
0                  2.5000  strong but artificial taste; peanut butter bet...  
1                  2.7500  too much stuffing to eat easily, cookies separ...  
2                  3.2500                     strong artificial lemon flavor  
3                  3.4000  weak chocolate flavor, tasted like an American...  
4                  3.4000  didn't taste like birthday cake, most like rea...  
5                  3.4000                                      default Oreos  
6                  3.6000  realistic red velvet taste with cream cheese f...  
7                  3.7000    tasted like coffee, specifically Dunkin' coffee  
8                  3.8125                                   taste like Oreos  
9                  4.0000  very light coconut flavor, more like a coconut...  
10                 4.0500     strong mint flavor, creme alone was toothpasty  
11                 4.3750  strong cinnamon flavor, did not capture the es...

**Part B:** In the remainder of this problem, you will perform some hypothesis tests to examine whether these data suggest there are significant differences in the mean ratings of the different Oreo flavors. Pick a level of significance for these experiments, and explain how you decided to use that significance level. "Because we used it a lot in class" is ***not*** a good reason.

**Solution:**

I like to pick $\alpha = 0.05$ significance level becuase I want to be $95\%$ confident that my analysis is correct and $5\%$ level is probability of rejection the null hypothesis when it's true.

As we know, aplha level is the probalility of making a Type I error, it seems to make sense that we make this area as small as possible. For instance, if we we the alpha level at $10\%$, then there is large chance that we might not correctly reject the null hypothesis, while an alpha level of $1\%$ would make the area too small. so I don't use $1\%$ instead of $5\%$ because the smaller alpha level, the smaller area that we would reject the null, when in fact we should. This's called Type II error.

The more we try and avoid a Type I error, the move likely a Type II error could creep in. Thus, I think alphal level of $5\%$ is a good balance between these two issues.



**Part C:** Perform an **Analysis of Variance hypothesis test** in order to determine if there is evidence that there is _some_ difference among the mean ratings given to these 12 Oreo flavors. Clearly state your null and alternative hypothesis, and use the significance level identified in **Part B**. You must show **all** calculations **by hand** (and may of course use Python as a calculator, and to compute values from a distribution using the appropriate percent-point-function (ppf) or cumulative distribution function (cdf)).

In addition to showing the code for your calculations, make comments **in Markdown** explaining what you are doing.

**Solution:**

Hypothese of interest in ANOVA are as follow:

$$
\begin{array}{ll}
H_0: & \mu_1 = \mu_2 = \ldots = \mu_{k} = 0  \\
H_1: & \mu_k \neq 0 \textrm{ for at least one } k=1, \ldots, 11  \\
\end{array}
$$
Where k are the number of independent group

Using F-test:
$$ F = \dfrac{SSB/SSB_{df}}{SSW/SSW_{df}}$$

where, $SSB = \sum_{i=0}^I n_i (\bar{y}_i - \bar{\bar{y}})^2$

$SSW = \sum_{i=0}^I \sum_{j=0}^{n_i} (y_{ij} - \bar{y}_i)^2 $

**Rejection region:** 

$F \ge F_{\alpha,p,n-p-1}$ = stats.f.ppf(1-$\alpha$, dfn = I-1, dfd= N-I)

**p-value:** 1 - stats.f.cdf(F,p,n-p-1) < $\alpha$


In [3]:
dfnew1 = df.copy()
dfnew1 = dfnew1.replace({"-":0})
dfOreo = pd.DataFrame(columns=['pb&j', 'mega stuf', "lemon", "chocolate","birthday cake","double stuf",
                               "red velvet","Dunkin' Donuts Mocha","mini","thins, coconut","mint","cinnamon bun"])

list_ = {0:'pb&j', 1:'mega stuf', 2:"lemon", 3:"chocolate",4:"birthday cake",5:"double stuf",
            6:"red velvet",7:"Dunkin' Donuts Mocha",8:"mini",9:"thins, coconut",10:"mint",11:"cinnamon bun"}

def newDataFrame(df, list_):
    df_new = pd.DataFrame(columns=['pb&j', 'mega stuf', "lemon", "chocolate","birthday cake","double stuf",
                               "red velvet","Dunkin' Donuts Mocha","mini","thins, coconut","mint","cinnamon bun"])
    
    for l in list_:
        x = df.loc[df['oreo_flavor'] == list_[l], 'rater_n':'rater_a']
        x = pd.to_numeric(x.loc[l])
        df_new[list_[l]] = x
    return df_new

dfOreo = newDataFrame(dfnew1, list_)
# dfOreo = dfOreo.dropna()
dfOreo

pb&j  mega stuf  lemon  chocolate  birthday cake  double stuf  \
rater_n   1.5        2.5    4.5        3.5            2.5          3.5   
rater_c   4.0        3.0    3.5        3.0            5.0          4.0   
rater_r   2.0        4.0    1.5        3.5            2.5          3.0   
rater_k   0.0        0.0    0.0        3.0            4.0          3.0   
rater_a   0.0        1.5    3.5        4.0            3.0          3.5   

         red velvet  Dunkin' Donuts Mocha  mini  thins, coconut  mint  \
rater_n         4.0                   2.5  3.00             3.5  3.50   
rater_c         3.0                   1.0  4.00             4.5  4.00   
rater_r         4.5                   5.0  4.00             3.0  4.50   
rater_k         4.0                   5.0  0.00             4.5  4.00   
rater_a         2.5                   5.0  4.25             4.5  4.25   

         cinnamon bun  
rater_n           0.0  
rater_c           4.5  
rater_r           4.5  
rater_k           4.5  
rater_a           4.0

In [4]:
f_alp = stats.f.ppf(1-0.05, 11, 54-12)
print(f_alp)

2.0252288250763315


In [5]:
# Get total number of points
N = 54
print(N)

# Get total number of groups
I = 12
print(I)

# Compute the grand_mean
grand_mean = np.mean(dfOreo.values) 

# Compute the between-group sum of squares 
SSB = np.sum([dfOreo[col].count()*(dfOreo[col].mean()-grand_mean)**2 for col in dfOreo.columns])

# Compute the between_group degrees of freedom 
SSB_df = I-1 

# Compute the within-group sum of squares 
SSW = np.sum([np.sum((dfOreo[col] - dfOreo[col].mean())**2) for col in dfOreo.columns])

# Compute the within_group degrees of freedom 
SSW_df = N-I 

# Compute the test statistic 
F = (SSB/SSB_df)/(SSW/SSW_df) 

# Compute the associated p-value 
pval = 1 - stats.f.cdf(F, SSB_df, SSW_df) 

print("SSB, SSB_df = {:.3f}, {}".format(SSB, SSB_df))
print("SSW, SSW_df = {:.3f}, {}".format(SSW, SSW_df))
print("F, pval = {:.5f}, {:.5f}".format(F, pval))

if F >= f_alp and pval < 0.05:
    print("Reject the null hypothesis.")
else: 
    print("Fail to reject the null hypothesis.")

54
12
SSB, SSB_df = 31.275, 11
SSW, SSW_df = 86.700, 42
F, pval = 1.37732, 0.21921
Fail to reject the null hypothesis.


**Rejection Rejion:**
Since $F = 1.377 < 2.025$, we fail to reject the null hypothesis. Therefore, performing at $95\%$ confidence level, we don't statistically sufficient evidence to claim that there is some difference among the mean ratings given to these 12 Oreo flavors.

**P-value:**
Since p-value $= 0.219 > 0.05$, we fail to reject the null hypothesis. Therfore, performing at $95\%$ confidence level, we don't have statistically sufficient evidence to claim that there is some difference among the mean ratings given to these 12 Oreo flavors.

**Part D:** Perform a **t hypothesis test** to determine if there is evidence supporting the claim that cinnamon bun Oreos have a higher mean rating than PB&J Oreos. Use the significance level you identified in **Part B**. Clearly state your null and alternative hypotheses, your conclusions, and show all work. Again, you may not use any canned t-test function.

**Solution:**
We will perform a hypothesis test for the difference between the means, we will use a t-test. 

For claiming that cinnamon bun Oreos have a higher mean rating than PB&J, we have hypothesis:

$$H_0 : \mu_{C} - \mu_{PB} = 0$$
$$H_1 : \mu_{C} - \mu_{PB} < 0 $$

The test-statistic is given :
$$T = \dfrac{\bar{x}_C - \bar{x}_{PB} }{\sqrt{\frac{ s_{C}^2 }{n_C} + \frac{ s_{PB}^2 }{n_{PB}} } }$$

**Rejection region:** $T < - t_{\alpha, n-2} = -2.015$

**P-value Test**, p-value = 1-stats.t.cdf(T, 22) $< \alpha$

I will use the original data set for calculation.

In [6]:
t_alp = stats.t.ppf(0.05, 7-2)
print("t_alpha = {:.03f}".format(t_alp))


t_alpha = -2.015


In [7]:
dfpb = df.copy()
# xc = dfpb.loc[dfpb['oreo_flavor'] == 'cinnamon bun', 'rater_c':'rater_a']
# xc = pd.to_numeric(xc.loc[11])
xc = dfpb.loc[dfpb['oreo_flavor'] == 'cinnamon bun', 'rater_c':'rater_a']
xc = pd.to_numeric(xc.loc[11])
alpha = 0.05
xbar_C = 4.375
nC = 4
S_C_sqrt = np.sum((xc - xbar_C)**2)/(nC-1)
# print("xc ", xc)
# print("cinnamon ", S_C_sqrt)
dfc = df.copy()
x_pb = dfc.loc[dfc['oreo_flavor'] == 'pb&j', 'rater_n':'rater_r']
# print(x_pb)
x_pb = pd.to_numeric(x_pb.loc[0])
xbar_PB = 2.5
n_PB = 3
S_PB_sqrt = np.sum((x_pb - xbar_PB)**2)/(n_PB-1)
# print("x_pb ", x_pb)
# print("P&b&j ", S_PB_sqrt)


dof = nC + n_PB - 2


numerator = xbar_C - xbar_PB
denomerator = np.sqrt((S_C_sqrt)/nC + (S_PB_sqrt)/n_PB)
T = numerator / denomerator

# print(t_alp)
print("Test_statistic T = {:.3f}".format(T))
p_value =stats.t.cdf(T,22)
print("p-value = {:.3f}".format(p_value))
if T < t_alp  and p_value < 0.02:
    print("Reject Null Hypothesis")
else:
    print("Failed to reject Null")

Test_statistic T = 2.423
p-value = 0.988
Failed to reject Null


**Rejection Rejion:**
Since $T = 2.423 > -2.015$, we fail to reject the null hypothesis. Therefore, performing at $95\%$ confidence level, we don't statistically sufficient evidence to claim that cinnamon bun Oreos have a higher mean rating than PB&J.

**P-value:**
Since p-value = 0.988 > 0.05, we fail to reject the null hypothesis. Therfore, performing at $95\%$ confidence level, we don't have statistically sufficient evidence to claim that cinnamon bun Oreos have a higher mean rating than PB&J.

**Part E:** Do your results from Parts C and D agree with one another? If they agree, explain how they are in agreement in _words_. If they do not agree, explain why you think they do not agree.

_Hint: if they do not agree, consider carefully what assumptions we make in ANOVA and the t-test._

**Solution:**

The result from Parts C and D agree with one another because the p-value is greater than significant level, the F-test is less than the f_alphat and th t-test is greater than the t_alpha, so we fail to reject Null Hypothesis. Therefore, performing at 95% confidence level, we don't have statistically sufficient evidence to claim that there is some difference among the mean ratings given to these 12 Oreo flavors.
    

<br>

---
<a id='p2'></a>
[Back to top](#top)

### [30 points] Problem 2: Multiple Linear Regression to Explain House Hauntings

<img src="https://s-media-cache-ak0.pinimg.com/originals/09/72/01/09720128cff5de4d4af038cd3fcf7f69.jpg" style="width: 300px;"/>

In an effort to control the skyrocketing prices of real estate in the Colorado Front Range, Governor Polis implemented a cutting edge new intervention. This new program oversaw the introduction of ghosts back into their natural ecosystem, after the ghost population seriously dwindled in recent decades due to overhaunting. However, an unfortunate miscalculation has led to haunted houses becoming a very serious problem in Colorado. Modern problems require modern solutions, so Governor Polis has hired you and the famous hedgehog data scientist/part-time ghostbuster Amy to determine what features of a house may be used to best predict a `haunted` score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted).

You decide to use multiple linear regression to understand and predict what factors lead to increased haunted house hazard. You collected a data set from Haunted Zillow, the lesser-known database of haunted house prices and attributes. The data cover a variety of potential features, and you'll find this data in the file `houses.csv`. 

**Response**: 

- $\texttt{haunted}$: a haunting score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted)

**Features**: 

- $\texttt{age}$: age of the house, in years
- $\texttt{area}$: square footage of interior of house
- $\texttt{bathrooms}$: number of bathrooms
- $\texttt{distance metro}$: distance to the nearest major metropolitan area (in miles)
- $\texttt{distance cemetery}$: distance to the nearest cemetery (in miles)
- $\texttt{cats}$: the number of cats within a one-block radius of the house
- $\texttt{howls}$: the number of wolf howls heard on an average night in the house's neighborhood
- $\texttt{clouds}$: what percentage of the sky was covered by clouds (fraction, 0-1)
- $\texttt{precipitation}$: amount of precipitation in the past 72 hours (inches)
- $\texttt{misery index}$: an economic indicator for how miserable the average United States citizen is, based on the unemployment rate and the inflation rate. More [here](https://www.stuffyoushouldknow.com/podcasts/whats-the-misery-index.htm) and [here](https://en.wikipedia.org/wiki/Misery_index_(economics)). Higher values correspond to more miserable citizens.
- $\texttt{ice cream sold}$: the number of units of ice cream sold at the farmer's market the week the house was most recently sold

**Part A**: Read the data from `houses.csv` into a Pandas DataFrame.  Note that since we will be doing a multiple linear regression we will need all of the features, so you should drop any row in the DataFrame that is missing data. 

In [8]:
dfHouse = pd.read_csv("houses.csv")
dfHouse = dfHouse.dropna()
dfHouse.head()

age  area  bathrooms  distance metro  distance cemetery  cats  howls  \
0   65.06  2041        1.0             7.1              10.01     7      3   
1  141.48  1564        0.0             7.4               4.07     5      5   
2    7.58  1637        3.0             7.0               3.36     2      0   
3   51.47  2021        2.0             7.9               3.43     6      8   
4  259.44  1642        1.0             7.5               3.19     4      1   

   clouds  precipitation  misery index  ice cream sold   haunted  
0    1.00           0.82         12.99             273 -0.596150  
1    1.00           0.99         16.77             184 -0.146465  
2    1.00           1.17         16.49             141 -0.303117  
3    0.13           0.92          8.28             146  0.339912  
4    1.00           1.73          5.90             178  0.724867

**Part B**: Perform the appropriate statistical test at the $\alpha = 0.01$ significance level to determine if _at least one_ of the features is related to the the response $y$.  Clearly describe your methodology and show all computations in Python. 

**Solution:**

We perform F-test with hypotheses:

$$
\begin{array}{ll}
H_0: & \beta_1 = \beta_2 = \ldots = \beta_{11} = 0  \\
H_1: & \beta_i \neq 0 \textrm{ for at least one } i=1, \ldots, 11  \\
\end{array}
$$

We use sm.OLS to perform this test. We have

In [9]:
y = dfHouse["haunted"]
X = dfHouse.loc[:,dfHouse.columns != "haunted"]
X = sm.add_constant(X)
full_model = sm.OLS(y, X).fit()
full_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                haunted   R-squared:                       0.948
Model:                            OLS   Adj. R-squared:                  0.938
Method:                 Least Squares   F-statistic:                     94.60
Date:                Thu, 02 May 2019   Prob (F-statistic):           1.97e-32
Time:                        13:45:40   Log-Likelihood:                 13.575
No. Observations:                  69   AIC:                            -3.151
Df Residuals:                      57   BIC:                             23.66
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.3480      0.323      1.078      0.285      -0.298       0.994
age                   0.0017      0.000      5.859      0.000       0.001       0.002
area                5.83e-05   6.16e-05      0.946      0.348   -6.51e-05       0.000
bathrooms            -0.0411      0.023     -1.806      0.076      -0.087       0.004
distance metro       -0.0142      0.032     -0.443      0.660      -0.079       0.050
distance cemetery    -0.1000      0.003    -30.867      0.000      -0.106      -0.093
cats                  0.0316      0.010      3.258      0.002       0.012       0.051
howls                 0.0056      0.011      0.494      0.623      -0.017       0.029
clouds                0.0902      0.123      0.734      0.466      -0.156       0.336
precipitation         0.0068      0.067      0.102      0.919      -0.128       0.141
misery index         -0.0082      0.006     -1.456      0.151      -0.020       0.003
ice cream sold       -0.0006      0.001     -1.274      0.208      -0.002       0.000
==============================================================================
Omnibus:                        1.629   Durbin-Watson:                   2.191
Prob(Omnibus):                  0.443   Jarque-Bera (JB):                1.480
Skew:                           0.230   Prob(JB):                        0.477
Kurtosis:                       2.450   Cond. No.                     2.55e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Conclusion:**

From the model summery, we see F = 94.60 and p-value $1.94 \times 10^{-32}$.
Since the p-value is less than 𝛼=0.01 we reject the null hypothesis and conclude that there is the sufficient statistical evidence that $\beta \ne 0$ and conclude that at least one of the provided features is related to the response.

**Part C**: Write a function `forward_select(df, resp_str, maxk)` that takes in the DataFrame, the name of the column corresponding to the response, and the maximum number of desired features, and returns a list of feature names corresponding to the `maxk` most important features via forward selection.  At each stage in forward selection you should add the feature whose inclusion in the model would result in the lowest sum of squared errors $(SSE)$. Use your function to determine the best $k=5$ features to include in the model. Clearly indicate which feature was added in each stage. 

**Note**: The point of this exercise is to see if you can implement **foward_select** yourself.  You may of course use canned routines like statmodels OLS, but you may not call any Python method that explicitly performs forward selection.

In [10]:
def forward_select(df, resp_str, maxk=5):
    y = df[resp_str]
    features= list(df.columns[df.columns != resp_str])
    keep = ""
    included = []
    stage = 0
    while stage < maxk:
        min_ = 100000
        for new_column in features:
            X = df[new_column]
            X = sm.add_constant(X)
            model = sm.OLS(y, X).fit()
            sse = np.sum((y - model.predict(X))**2)
#             print("sse before: {}, {} ".format(sse, new_column))
            if sse < min_:
                min_ = sse
                keep = new_column

        included.append(keep)
        features.remove(keep)
        stage +=1
        print("stage {}, add '{}' with SSE = {:.3f}".format(stage, keep, min_))
    print("Features add:",included)
    return included
desired_features = forward_select(dfHouse, "haunted", maxk=5)

stage 1, add 'distance cemetery' with SSE = 5.789
stage 2, add 'ice cream sold' with SSE = 50.893
stage 3, add 'clouds' with SSE = 51.644
stage 4, add 'distance metro' with SSE = 51.678
stage 5, add 'bathrooms' with SSE = 51.914
Features add: ['distance cemetery', 'ice cream sold', 'clouds', 'distance metro', 'bathrooms']


**Part D**: Write down the multiple linear regression model, including estimated parameters, obtained by your forward selection process. 

In [11]:
y = dfHouse["haunted"]
Xreduce = dfHouse.loc[:, desired_features]
Xreduce = sm.add_constant(Xreduce)
reduce_model = sm.OLS(y, Xreduce).fit()
reduce_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                haunted   R-squared:                       0.902
Model:                            OLS   Adj. R-squared:                  0.894
Method:                 Least Squares   F-statistic:                     115.7
Date:                Thu, 02 May 2019   Prob (F-statistic):           2.18e-30
Time:                        13:45:41   Log-Likelihood:                -8.4119
No. Observations:                  69   AIC:                             28.82
Df Residuals:                      63   BIC:                             42.23
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 1.1294      0.310      3.647      0.001       0.511       1.748
distance cemetery    -0.0950      0.004    -23.122      0.000      -0.103      -0.087
ice cream sold       -0.0013      0.001     -1.976      0.053      -0.003     1.4e-05
clouds                0.1117      0.156      0.715      0.477      -0.201       0.424
distance metro       -0.0451      0.040     -1.121      0.266      -0.126       0.035
bathrooms            -0.0525      0.029     -1.811      0.075      -0.110       0.005
==============================================================================
Omnibus:                        0.218   Durbin-Watson:                   1.984
Prob(Omnibus):                  0.897   Jarque-Bera (JB):                0.245
Skew:                          -0.125   Prob(JB):                        0.885
Kurtosis:                       2.850   Cond. No.                     1.81e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.81e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Solution:** The multiple linear regression from the forward selection is :

$$\hat{y} = 1.1294 + (-0.095) \times \text{distance cemetery} + (-0.0013) \times \text{ice creams sold} + (0.1117) \times \text{clouds}+(-0.0451) \times \text{distance metro} + (-0.0525) \times \text{bathrooms}$$

**Part E**: Perform the appropriate statistical test at the $\alpha = 0.05$ significance level to determine whether there is a statistically significant difference between the full model with all features and the reduced model obtained by forward selection in **Part D**. 

**Solution:**

$$
\begin{array}{ll}
H_0: \beta_{age} = \beta_{area}= \beta_{cats} = \beta_{howls}  = \beta_{precipitation} = \beta_{misery index} = 0 \\
H_1: \beta_{k} \neq 0 \textrm{ for at least one } k \in \{age,area, cats, howls, precipitation, misery index\} \\
\end{array}
$$

We will calcualte a Partial F-Test of the form:

$$
F = \dfrac{(SSE_{reduce} - SSE_{full})/(p-k)}{SSE_{full}/(n-p-1)}
$$

**Rejection region:** 

$F \ge F_{\alpha,p,n-p-1}$ = stats.f.ppf(1-$\alpha$, dfn = I-1, dfd= N-I)

**p-value:** 1 - stats.f.cdf(F,p,n-p-1) < $\alpha$


In [12]:
#full model
y_full = dfHouse["haunted"]
X_full = dfHouse.loc[:,dfHouse.columns != "haunted"]
X_full = sm.add_constant(X_full)
fullModel = sm.OLS(y_full, X_full).fit()
yhat_full = fullModel.predict(X_full)
SSE_full = np.sum((y_full - yhat_full)**2)
print("SSE full = {:.03f}".format(SSE_full))

#reduce model
y_reduce = dfHouse["haunted"]
Xreduce = dfHouse.loc[:, desired_features]
Xreduce = sm.add_constant(Xreduce)
reduceModel = sm.OLS(y_reduce, Xreduce).fit()
yhat_reduce = reduceModel.predict(Xreduce)
SSE_reduce = np.sum((y_reduce - yhat_reduce)**2)
print("SSE reduce = {:.03f}".format(SSE_reduce))
n = len(dfHouse)
p = len(dfHouse.columns) - 1
k = len(desired_features)

numerator = (SSE_reduce - SSE_full)/(p-k)
denomerator = SSE_full/(n-p-1)

F = numerator/denomerator

print("F = {:.05f}".format(F))
pvalue = 1 - stats.f.cdf(F, p-k, n-p-1)
print("p-value = {}".format(pvalue))
f_alp = stats.f.ppf(1-0.04, p-k, n-p-1)
print("F alphal = {:.05f}".format(f_alp))

if F >= f_alp and pvalue < 0.05:
    print("Reject the null hypothesis.")
else: 
    print("Fail to reject the null hypothesis.")

SSE full = 2.726
SSE reduce = 5.155
F = 8.46826
p-value = 1.3942144432288472e-06
F alphal = 2.38421
Reject the null hypothesis.


**Conclution:**

Since p-value $ < \alpha = 0.05$, we reject the null hypothesis. Therefore, we conclude that there is the sufficient evident to claim that the full model is statistically significantly better than the reduced model obtained by forward selection.

**Part F**: Based on your conclusions in **Part E**, use the _better_ of the two models to predict the sharknado hazard when the following features are observed: 

- $\texttt{age}$: 100 years
- $\texttt{area}$: 2200 square feet
- $\texttt{bathrooms}$: 3 bathrooms
- $\texttt{distance metro}$: 25 miles
- $\texttt{distance cemetery}$: 10 miles
- $\texttt{cats}$: 4 cats
- $\texttt{howls}$: 5 wolf howls/night
- $\texttt{clouds}$: 0.65 cloud cover
- $\texttt{precipitation}$: 0 inches
- $\texttt{misery index}$: 10
- $\texttt{ice cream sold}$: 125

In [13]:
newData = {"const" : 1, "age" : 100, "area" : 2200, "bathrooms" : 3, "distance metro" : 25, \
          "distance cemetery" : 10, "cats": 4, "howls":5, "clouds" : 0.65, "precipitation" : 0, \
          "misery index" : 10, "ice cream sold" : 125}


yhat = 0
for f, slope in zip(reduceModel.params.index, reduceModel.params):
    yhat += newData[f] * slope 
print("yhat = {:.3f}".format(yhat))




yhat = -1.191


**Part G:** Governor Polis dabbles a bit in the art of data science, as well as the science of data art. He tells you that the response (`haunted` score) that you and Amy predicted is actually the natural logarithm of the _odds_ that a house with the given features is haunted, where if $p$ is the probability that a house is haunted, then the odds are given by $$\text{odds} = \dfrac{p}{1-p}$$

What transformation must you make in order to turn your multiple regression model into a logistic regression model, to classify a house a haunted or not? Perform this simple transformation, then use a decision threshold of 0.5 to classify the house from **Part F** as haunted or not haunted. No new models should be fit here; use the same model that you used in Part F.

In [14]:
def sigm(z):
    return 1 / (1 + np.exp(-z))

P = sigm(yhat)
odd = p/(1-P)
print("P: {}".format(P))
if P > 0.5 :
    print("The house classify as haunted")
else:
    print("The house classify as Not haunted")

P: 0.23303444702599865
The house classify as Not haunted


The house is not haunted because it's less than threshold of 0.5.

<br>

---
[Back to top](#top)
<a id='bot'></a>